cEP0002.md

```diff
@@ -0,0 +1,272 @@
+        self.file_dict = file_dict
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def execute_task(self, task, **kwargs):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        for result in self.run(task, **self.kwargs):
+            # This queue holds the results locally for a bear.
+            self.result_queue.put(result)
+
+    def run(self,
+            file_proxy,
+            *args,
+            dependency_results=None,
+            **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        for fileproxy in self.file_dict.values():
+            yield fileproxy
+```
+
+A test bear:
+
+```python
+class TestBear(LocalBear):
+    def run(self, file_proxy):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        for i, line in enumerate(file):
+            # This will contain the analysis routine
+            yield "Line has a problem" , file_proxy.name, i
+```
+
+A rough idea of how the new bear will be called:
+
+```python
+from multiprocessing import Pool, Queue
+from queue import Empty
+
+
+def all_results_ready(async_results):
+    """
+    Method checks if all bears have completed.
+    :param async_results: A dictionary with bears as keys result objects as
+                          values.
+    """
+    for i, result in reversed(enumerate(async_results)):
+        if result.ready():
+            del async_results[i]
+        else:
+            return False
+
+    return True
+
+
+def runstuff(handle_result):
+    queue = Queue()
+    bears = [LocalBear({}, queue, {})]
+    dep_tracker = Dependencies()
+
+    pool = Pool()
+    results = []
+    async_results = {}
+    for bear in bears:
+        # This is how the multiprocessing pool will call the bear.
+        if bear.get_dependencies():
+            for dep_bear in bear.get_dependencies():
+                dep_tracker.add_dependency(bear, dep_bear)
+        else:
+            async_results[bear] = pool.map_async(
+                bear.__execute_task,
+                bear.generate_tasks())
+
+    while True:
+        try:
+            result = queue.get(0.1)
+            if isinstance(result, Result):
+                handle_result(result)
+                results.append(result)
+            elif isinstance(result, LogMessage):
+                logger.log(result)
```

Comment:
the bear would use `self.log` which does this under the hood. I think it simplifies handling a lot when communication goes over a single queue^^